shotover-proxy/src/transforms/redis/sink_single.rs

```diff
@@ -73,43 +86,230 @@ impl Transform for RedisSinkSingle {
+async fn server_response_processing_task(
+    mut outbound_rx: SplitStream<RedisFramed>,
+    subscribe_tx: Option<mpsc::UnboundedSender<Messages>>,
+    response_messages_tx: mpsc::UnboundedSender<Message>,
+    mut sent_message_type: mpsc::UnboundedReceiver<MessageType>,
+) {
+    let mut is_subscribed = true;
+    loop {
+        tokio::select! {
+            responses = outbound_rx.next().fuse() => {
+                if process_server_response(
+                    responses,
+                    &subscribe_tx,
+                    &response_messages_tx,
+                    &mut is_subscribed,
+                    &mut sent_message_type
+                ).await {
+                    return;
+                }
+            },
+            _ = response_messages_tx.closed() => {
+                tracing::debug!("RedisSinkSingle dropped, redis single subscription task shutting down");
+                return;
+            },
+        }
+    }
+}
+
+/// returns true when the task should shutdown
+async fn process_server_response(
+    responses: Option<Result<Messages>>,
+    subscribe_tx: &Option<mpsc::UnboundedSender<Messages>>,
+    response_messages_tx: &mpsc::UnboundedSender<Message>,
+    is_subscribed: &mut bool,
+    sent_message_type: &mut mpsc::UnboundedReceiver<MessageType>,
+) -> bool {
+    match responses {
+        Some(Ok(messages)) => {
+            for mut message in messages {
+                // Notes on subscription responses
+                //
+                // There are 3 types of pubsub responses and the type is determined by the first value in the array:
+                // * `subscribe` - a response to a SUBSCRIBE, PSUBSCRIBE or SSUBSCRIBE request
+                // * `unsubscribe` - a response to an UNSUBSCRIBE, PUNSUBSCRIBE or SUNSUBSCRIBE request
+                // * `message` - a subscription message
+                //
+                // Additionally redis will:
+                // * accept a few regular commands while in pubsub mode: PING, RESET and QUIT
+                // * return an error response when a nonexistent or non pubsub compatible command is used
+                //
+                // Note: PING has a custom response when in pubsub mode.
+                //       It returns an array ['pong', $pingMessage] instead of directly returning $pingMessage.
+                //       But this doesnt cause any problems for us.
+
+                // Determine if message is a `message` subscription message
+                //
+                // Because PING, RESET, QUIT and error responses never return a RedisFrame::Array starting with `message`,
+                // they have no way to collide with the `message` value of a subscription message.
+                // So while we are in subscription mode we can use that to determine if an
+                // incoming message is a subscription message.
+                let is_subscription_message = if *is_subscribed {
```

Comment:
Can simplify to: 

```rust
let is_subscription_message = {
    if *is_subscribed {
        if let Some(Frame::Redis(RedisFrame::Array(array))) = message.frame() {
            if let [RedisFrame::BulkString(ty), ..] = array.as_slice() {
                return ty.as_ref() == b"message";
             }
        }
    }
    false
};
```

Reply:
are you implying for me to pull it out into a separate function?
Otherwise that return wont work

Reply:
Oh, my mistake. I thought the return would be scoped.

Reply:
weird.
When I make that same change it fails with:
```
---- redis_int_tests::basic_driver_tests::test_passthrough stdout ----
thread 'tokio-runtime-worker' panicked at 'called `Option::unwrap()` on a `None` value', shotover-proxy/src/transforms/redis/sink_single.rs:153:76
stack backtrace:
   0: rust_begin_unwind
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/panicking.rs:142:14
   2: core::panicking::panic
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/panicking.rs:48:5
   3: core::option::Option<T>::unwrap
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/option.rs:755:21
   4: <shotover_proxy::transforms::redis::sink_single::RedisSinkSingle as shotover_proxy::transforms::Transform>::transform::{{closure}}
             at ./src/transforms/redis/sink_single.rs:153:31
   5: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/future/mod.rs:91:19
   6: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/future/future.rs:124:9
   7: shotover_proxy::transforms::Transforms::transform::{{closure}}
             at ./src/transforms/mod.rs:123:75
   8: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/future/mod.rs:91:19
   9: <tokio::task::task_local::TaskLocalFuture<T,F> as core::future::future::Future>::poll::{{closure}}
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/task/task_local.rs:280:28
  10: tokio::task::task_local::TaskLocalFuture<T,F>::with_task
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/task/task_local.rs:272:9
  11: <tokio::task::task_local::TaskLocalFuture<T,F> as core::future::future::Future>::poll
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/task/task_local.rs:280:9
  12: shotover_proxy::transforms::Wrapper::call_next_transform::{{closure}}
             at ./src/transforms/mod.rs:399:58
  13: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/future/mod.rs:91:19
  14: shotover_proxy::transforms::chain::TransformChain::process_request::{{closure}}
             at ./src/transforms/chain.rs:279:51
  15: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/future/mod.rs:91:19
  16: shotover_proxy::server::Handler<C>::run::{{closure}}
             at ./src/server.rs:498:75
  17: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/future/mod.rs:91:19
  18: shotover_proxy::server::TcpCodecListener<C>::run::{{closure}}::{{closure}}
             at ./src/server.rs:247:78
  19: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/future/mod.rs:91:19
  20: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-0.1.35/src/instrument.rs:272:9
  21: tokio::runtime::task::core::CoreStage<T>::poll::{{closure}}
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/task/core.rs:165:17
  22: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/loom/std/unsafe_cell.rs:14:9
  23: tokio::runtime::task::core::CoreStage<T>::poll
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/task/core.rs:155:13
  24: tokio::runtime::task::harness::poll_future::{{closure}}
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/task/harness.rs:470:19
  25: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/panic/unwind_safe.rs:271:9
  26: std::panicking::try::do_call
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/panicking.rs:492:40
  27: __rust_try
  28: std::panicking::try
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/panicking.rs:456:19
  29: std::panic::catch_unwind
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/panic.rs:137:14
  30: tokio::runtime::task::harness::poll_future
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/task/harness.rs:458:18
  31: tokio::runtime::task::harness::Harness<T,S>::poll_inner
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/task/harness.rs:104:27
  32: tokio::runtime::task::harness::Harness<T,S>::poll
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/task/harness.rs:57:15
  33: tokio::runtime::task::raw::poll
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/task/raw.rs:144:5
  34: tokio::runtime::task::raw::RawTask::poll
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/task/raw.rs:84:18
  35: tokio::runtime::task::LocalNotified<S>::run
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/task/mod.rs:376:9
  36: tokio::runtime::thread_pool::worker::Context::run_task::{{closure}}
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/thread_pool/worker.rs:458:21
  37: tokio::coop::with_budget::{{closure}}
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/coop.rs:102:9
  38: std::thread::local::LocalKey<T>::try_with
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/thread/local.rs:445:16
  39: std::thread::local::LocalKey<T>::with
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/thread/local.rs:421:9
  40: tokio::coop::with_budget
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/coop.rs:95:5
  41: tokio::coop::budget
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/coop.rs:72:5
  42: tokio::runtime::thread_pool::worker::Context::run_task
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/thread_pool/worker.rs:434:9
  43: tokio::runtime::thread_pool::worker::Context::run
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/thread_pool/worker.rs:401:24
  44: tokio::runtime::thread_pool::worker::run::{{closure}}
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/thread_pool/worker.rs:386:17
  45: tokio::macros::scoped_tls::ScopedKey<T>::set
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/macros/scoped_tls.rs:61:9
  46: tokio::runtime::thread_pool::worker::run
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/thread_pool/worker.rs:383:5
  47: tokio::runtime::thread_pool::worker::Launch::launch::{{closure}}
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/thread_pool/worker.rs:362:45
  48: <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/blocking/task.rs:42:21
  49: tokio::runtime::task::core::CoreStage<T>::poll::{{closure}}
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/task/core.rs:165:17
  50: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/loom/std/unsafe_cell.rs:14:9
  51: tokio::runtime::task::core::CoreStage<T>::poll
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/task/core.rs:155:13
  52: tokio::runtime::task::harness::poll_future::{{closure}}
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/task/harness.rs:470:19
  53: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/panic/unwind_safe.rs:271:9
  54: std::panicking::try::do_call
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/panicking.rs:492:40
  55: __rust_try
  56: std::panicking::try
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/panicking.rs:456:19
  57: std::panic::catch_unwind
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/panic.rs:137:14
  58: tokio::runtime::task::harness::poll_future
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/task/harness.rs:458:18
  59: tokio::runtime::task::harness::Harness<T,S>::poll_inner
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/task/harness.rs:104:27
  60: tokio::runtime::task::harness::Harness<T,S>::poll
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/task/harness.rs:57:15
  61: tokio::runtime::task::raw::poll
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/task/raw.rs:144:5
  62: tokio::runtime::task::raw::RawTask::poll
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/task/raw.rs:84:18
  63: tokio::runtime::task::UnownedTask<S>::run
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/task/mod.rs:413:9
  64: tokio::runtime::blocking::pool::Task::run
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/blocking/pool.rs:91:9
  65: tokio::runtime::blocking::pool::Inner::run
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/blocking/pool.rs:325:17
  66: tokio::runtime::blocking::pool::Spawner::spawn_thread::{{closure}}
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/blocking/pool.rs:300:13
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
thread 'redis_int_tests::basic_driver_tests::test_passthrough' panicked at 'assertion failed: `(left == right)`
  left: `Err(timed out)`,
 right: `Ok(("foo", [98, 97, 114]))`', shotover-proxy/tests/redis_int_tests/basic_driver_tests.rs:30:5
stack backtrace:
   0: rust_begin_unwind
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/panicking.rs:142:14
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/panicking.rs:181:5
   4: lib::redis_int_tests::basic_driver_tests::test_args::{{closure}}
             at ./tests/redis_int_tests/basic_driver_tests.rs:30:5
   5: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/future/mod.rs:91:19
   6: lib::redis_int_tests::basic_driver_tests::run_all::{{closure}}
             at ./tests/redis_int_tests/basic_driver_tests.rs:1641:26
   7: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/future/mod.rs:91:19
   8: lib::redis_int_tests::basic_driver_tests::test_passthrough::{{closure}}::{{closure}}
             at ./tests/redis_int_tests/basic_driver_tests.rs:1314:62
   9: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/future/mod.rs:91:19
  10: tokio::park::thread::CachedParkThread::block_on::{{closure}}
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/park/thread.rs:263:54
  11: tokio::coop::with_budget::{{closure}}
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/coop.rs:102:9
  12: std::thread::local::LocalKey<T>::try_with
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/thread/local.rs:445:16
  13: std::thread::local::LocalKey<T>::with
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/thread/local.rs:421:9
  14: tokio::coop::with_budget
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/coop.rs:95:5
  15: tokio::coop::budget
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/coop.rs:72:5
  16: tokio::park::thread::CachedParkThread::block_on
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/park/thread.rs:263:31
  17: tokio::runtime::enter::Enter::block_on
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/enter.rs:151:13
  18: tokio::runtime::thread_pool::ThreadPool::block_on
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/thread_pool/mod.rs:90:9
  19: tokio::runtime::Runtime::block_on
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/mod.rs:482:43
  20: lib::redis_int_tests::basic_driver_tests::test_passthrough::{{closure}}
             at ./tests/redis_int_tests/basic_driver_tests.rs:1314:5
  21: core::ops::function::FnOnce::call_once
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/ops/function.rs:248:5
  22: serial_test::serial_code_lock::local_serial_core
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/serial_test-0.8.0/src/serial_code_lock.rs:26:5
  23: lib::redis_int_tests::basic_driver_tests::test_passthrough
             at ./tests/redis_int_tests/basic_driver_tests.rs:1305:1
  24: lib::redis_int_tests::basic_driver_tests::test_passthrough::{{closure}}
             at ./tests/redis_int_tests/basic_driver_tests.rs:1306:7
  25: core::ops::function::FnOnce::call_once
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/ops/function.rs:248:5
  26: core::ops::function::FnOnce::call_once
             at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/ops/function.rs:248:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```